cmd/local-gadget/trace/dns.go

```diff
@@ -0,0 +1,156 @@
+	containercollection "github.com/inspektor-gadget/inspektor-gadget/pkg/container-collection"
+	dnsTracer "github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/dns/tracer"
+	dnsTypes "github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/dns/types"
+	localgadgetmanager "github.com/inspektor-gadget/inspektor-gadget/pkg/local-gadget-manager"
+	eventtypes "github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+)
+
+const logalGadgetSubKey = "local-gadget-key"
+
+func newDNSCmd() *cobra.Command {
+	var commonFlags utils.CommonFlags
+
+	// The DNS gadget works in a different way than most gadgets: It
+	// attaches a new eBPF program to each container when it's
+	// created instead of using an eBPF map with the mount
+	// namespaces IDs to filter the events. For this reason we can't
+	// use the TraceGadget implementation here.
+	runCmd := func(*cobra.Command, []string) error {
+		localGadgetManager, err := localgadgetmanager.NewManager(commonFlags.RuntimeConfigs)
+		if err != nil {
+			return commonutils.WrapInErrManagerInit(err)
+		}
+		defer localGadgetManager.Close()
+
+		// local-gadget is designed to trace containers, hence enable this column
+		cols := dnsTypes.GetColumns()
+		col, _ := cols.GetColumn("container")
+		col.Visible = true
+
+		parser, err := commonutils.NewGadgetParserWithRuntimeInfo(&commonFlags.OutputConfig, cols)
+		if err != nil {
+			return commonutils.WrapInErrParserCreate(err)
+		}
+
+		eventCallback := func(event dnsTypes.Event) {
+			baseEvent := event.GetBaseEvent()
+			if baseEvent.Type != eventtypes.NORMAL {
+				commonutils.ManageSpecialEvent(baseEvent, commonFlags.Verbose)
+				return
+			}
+
+			// TODO: This enrich logic could be moved to the
+			// gadget itself to simplify this here and in
+			// pkg/gadget-collection/gadgets/trace/dns/gadget.go
+			container := localGadgetManager.GetContainer(event.Key)
+			if container != nil {
+				event.Namespace = container.Namespace
+				event.Pod = container.Podname
+				event.Container = container.Name
+			}
+
+			switch commonFlags.OutputMode {
+			case commonutils.OutputModeJSON:
+				b, err := json.Marshal(event)
+				if err != nil {
+					fmt.Fprint(os.Stderr, fmt.Sprint(commonutils.WrapInErrMarshalOutput(err)))
+					return
+				}
+
+				fmt.Println(string(b))
+			case commonutils.OutputModeColumns:
+				fallthrough
+			case commonutils.OutputModeCustomColumns:
+				fmt.Println(parser.TransformIntoColumns(&event))
+			default:
+				fmt.Fprint(os.Stderr, commonutils.WrapInErrOutputModeNotSupported(commonFlags.OutputMode))
+			}
+		}
+
+		tracer, err := dnsTracer.NewTracer()
+		if err != nil {
+			return fmt.Errorf("creating tracer: %w", err)
+		}
+		defer tracer.Close()
+
+		if commonFlags.OutputMode != commonutils.OutputModeJSON {
+			fmt.Println(parser.BuildColumnsHeader())
+		}
+
+		attachContainerFunc := func(container *containercollection.Container) {
+			err = tracer.Attach(container.ID, container.Pid, eventCallback)
```

Comment:
I wonder if we should use the "host" key for the host network namespace to not enrich those events, as it could be confusing. 

Consider the user has this pod running:

```bash
$ kubectl run --image=busybox mypod -- sh -c 'while true; do wget -q -O /dev/null -T 3 https://www.inspektor-gadget.io; sleep 3; done'
```

And user runs `local-gadget` filtering by a container that uses the host network namespace, they will see the events from the `mypod` container:

```bash
$ sudo local-gadget trace dns -c etcd
WARN[0000] Runtime enricher (cri-o): couldn't get current containers
CONTAINER                                                                            QR NAMESERVER                                    TYPE      QTYPE      NAME
etcd                                                                                 Q  10.96.0.10                                    HOST      AAAA       www.inspektor-gadget.io.
etcd                                                                                 Q  192.168.0.41                                  OUTGOING  AAAA       www.inspektor-gadget.io.
etcd                                                                                 R  192.168.0.41                                  HOST      AAAA       www.inspektor-gadget.io.
etcd                                                                                 R  10.96.0.10                                    OUTGOING  AAAA       www.inspektor-gadget.io.
etcd                                                                                 Q  10.96.0.10                                    HOST      A          www.inspektor-gadget.io.
etcd                                                                                 Q  192.168.0.41                                  OUTGOING  A          www.inspektor-gadget.io.
etcd                                                                                 R  192.168.0.41                                  HOST      A          www.inspektor-gadget.io.
etcd                                                                                 R  10.96.0.10                                    OUTGOING  A          www.inspektor-gadget.io.
```

Notice there are some packets with `OUTGOING` type. I think it is confusing, and users could think they are going out from the container they filtered. They don't know how the gadget works and all the details about the network namespace. Don't you think?

Same for the `kubeclt-gadget`. Even though not having a namespace/pod is an indicator, I'm not sure if it is enough.  

```bash
$ ./kubectl-gadget trace dns -n kube-system -p etcd-master
NODE                                            NAMESPACE                                       POD                                            QR NAMESERVER               TYPE      QTYPE      NAME
master                                                                                                                                         Q  10.96.0.10               HOST      AAAA       www.inspektor-gadget.io.
master                                                                                                                                         Q  192.168.0.41             OUTGOING  AAAA       www.inspektor-gadget.io.
master                                                                                                                                         R  192.168.0.41             HOST      AAAA       www.inspektor-gadget.io.
master                                                                                                                                         R  10.96.0.10               OUTGOING  AAAA       www.inspektor-gadget.io.
master                                                                                                                                         Q  10.96.0.10               HOST      A          www.inspektor-gadget.io.
master                                                                                                                                         Q  192.168.0.41             OUTGOING  A          www.inspektor-gadget.io.
master                                                                                                                                         R  192.168.0.41             HOST      A          www.inspektor-gadget.io.
master                                                                                                                                         R  10.96.0.10               OUTGOING  A          www.inspektor-gadget.io.
^C
Terminating...
```

I consider we need to make the output clearer. Maybe adding a column that indicates whether the packet was captured in the host network namespace and describes what it means in the documentation 🤔  ... Anyway, of course, we could manage this in another PR. 

Consider also the case we filter by a container/pod that shares the network namespace with another. In that case, how to indicate to the user that those could also be generated in another container/pod? 🤯

Reply:
> Consider also the case we filter by a container/pod that shares the network namespace with another

I think the only relevant case is to share the host network namespace, I don't think containers share a namespace under other conditions. 

Tracing a container that shares the host network namespace (`docker run --network=host --name test-container -it --rm busybox /bin/sh`) will show all the dns traffic in the host (and other containers sharing it). I'm not really sure how to handle this scenario, for now I added some logic to avoid printing a (wrong) container name.